internal/kms/kms.go

```diff
@@ -352,3 +355,38 @@ func (k *Kms) loadDek(ctx context.Context, scopeId string, purpose KeyPurpose, r
 
 	return multi, nil
 }
+
+// DerivedReaderReader returns a reader from which keys can be read, using the
+// given wrapper, reader length limit, salt and context info. Salt and info can
+// be nil.
+//
+// Example:
+//	reader, _ := NewDerivedReader(wrapper, userId, jobId)
+// 	key := ed25519.GenerateKey(reader)
+func NewDerivedReader(wrapper wrapping.Wrapper, lenLimit int64, salt, info []byte) (*io.LimitedReader, error) {
+	const op = "kms.NewDerivedReader"
+	if wrapper == nil {
+		return nil, errors.New(errors.InvalidParameter, op, "missing wrapper")
+	}
+	if lenLimit < 20 {
+		return nil, errors.New(errors.InvalidParameter, op, "lenLimit must be >= 20")
+	}
+	var aeadWrapper *aead.Wrapper
+	switch w := wrapper.(type) {
+	case *multiwrapper.MultiWrapper:
+		raw := w.WrapperForKeyID("__base__")
+		var ok bool
+		if aeadWrapper, ok = raw.(*aead.Wrapper); !ok {
+			return nil, errors.New(errors.InvalidParameter, op, "unexpected wrapper type from multiwrapper base")
+		}
+	case *aead.Wrapper:
+		aeadWrapper = w
+	default:
+		return nil, errors.New(errors.InvalidParameter, op, "unknown wrapper type")
```

Comment:
Should this return an encrypt error?

Reply:
I think, this is an invalid parameter, since we don't even get to attempt encryption.

Reply:
Yes agreed this should be InvalidParam